File: Coding Problems - Python/Arrays/20 Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashmap = {
            "(" : ")",
            "[" : "]",
            "{" : "}"
        }
        st = []

        for i in s:
            if i in hashmap:
                st.append(i)
            else:
                if st and hashmap[st[-1]] == i:

                    st.pop()
                else:
                    return 0
        if st:
            return 0
        return 1

File: Coding_Problems_Java/20 Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> pDict = new HashMap<>();
        pDict.put('(', ')');
        pDict.put('{', '}');
        pDict.put('[', ']');

        Stack<Character> stack = new Stack<>();

        for (char element : s.toCharArray()) {
            if (pDict.containsKey(element)) {
                stack.push(element);
            } else {
                if (stack.isEmpty()) {
                    return false;
                } else {
                    if (element == pDict.get(stack.peek())) {
                        stack.pop();
                    } else {
                        return false;
                    }
                }
            }
        }

        return stack.isEmpty();
    }
}
